text/0000-lifecycle-all.md

```diff
@@ -0,0 +1,109 @@
+# Meta
+[meta]: #meta
+- Name: Lifecycle All Binary
+- Start Date: 2020-01-21
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+### `/cnb/lifecycle/all`
+A new lifecycle binary with the name `lifecycle/all` will be included in each released lifecycle archive. Within a builder image, it will be found at the path `/cnb/lifecycle/all`. When invoked, it will run the following phases sequentially:
+
+* detector
+* analyzer
+* restorer
+* builder
+* exporter
+
+Each of these phases will continue to be available individually as provided by the current lifecycle.
+
+### Platform API 0.3
+The existence of `all` will constitute a non-breaking but substantive change to the Platform API, bringing us to Platform API 0.3 . Although platform will interpret this as a breaking change, we do not have a mechanism for indicating non-breaking platform API change pre 1.0. Bumping the platform API number is necessary so that platforms like `pack` know whether or no they can use this feature.
+
+# Motivation
+[motivation]: #motivation
+
+This new binary supports two main goals.
+
+### Goal 1: faster `pack` builds
+By creating a single container and invoking `/cnb/lifecycle/all` instead of creating a container per lifecycle phase, pack can shave approximate `6s` from the execution time of `pack build`.
+
+`pack` may not always want to do this (see https://github.com/buildpacks/rfcs/pull/43 for an explanation of why this is not always ideal). However, it could be very beneficial to users that are wiling to trust their builder images.
+
+### Goal 2: Easier integration of the lifecycle into other platforms and CI tools
+As members of the community experiment with incorporating CNBs into other platform and CI tools, the difficulty of invoking each lifecycle phase correctly and sequentially draws many to the easier solution of using `pack` to orchestrate the lifecycle. While `pack` can provide a simple way for these developers to implement these integrations, there may be performance drawbacks when compared with using the lifecycle directly. `/cnb/lifecycle/all` will provide an easier interface, lowering the barrier to using the lifecycle without pack, thus enabling platform developers choose between `pack` and `lifecycle` based solely upon the needs of their integration.
+
+# What it is
+[what-it-is]: #what-it-is
+
+### Usage
+
+`/cnb/lifecycle/all <image-name>`
```

Comment:
Currently, pack prepends lifecycle output with the step of the container builds. Such as: `===> DETECTING`. 

Are we expecting the `/lifecycle/all` binary prepend the output or will pack still detect the current step in the build?

Reply:
In this case, pack would have to be more clever when invoking `/lifecycle/all` to know which phase is running. I suspect the `all` binary might have to be better about logging which phase is currently being run.

Reply:
As a stop gap we could add the phase to the lifecycle logs. I think the better longterm solution is the structured logging RFC I have long been threatening to submit. This should be a good forcing function.

Reply:
+1 to structured logging